blog/buildFrontMatter.ts

```diff
@@ -0,0 +1,106 @@
+import path from "path";
+import fs from "fs";
+import fm from "front-matter";
+import type {
+  PostFrontMatter,
+  PostFrontMatterCollection,
+  PostFrontMatterWithSlug,
+} from "./models";
+import { POSTS_SOURCE_DIR } from "./paths";
+
+/**
+ * Confirms the file is markdown or MDX
+ * @param filename - the file name
+ */
+function isMarkdownFilename(filename: unknown): filename is "*md" | "*mdx" {
+  return (
+    typeof filename === "string" &&
+    (filename.endsWith("md") || filename.endsWith("mdx"))
+  );
+}
+
+/**
+ * Sorts our front matter collection with most recent first
+ * @param a - first group of `[slug, frontMatter]`
+ * @param b - second group of `[slug, frontMatter]`
+ */
+function reverseSort(a: PostFrontMatterWithSlug, b: PostFrontMatterWithSlug) {
+  const aDate = new Date(a.date);
+  const bDate = new Date(b.date);
+  return bDate.getTime() - aDate.getTime();
+}
+
+function validateFrontMatter(frontMatter: PostFrontMatter): never | void {
+  const hasTitle = !!frontMatter.title;
+  const hasAuthor = !!frontMatter.author;
+  const hasDate = !!frontMatter.date;
+  const isValid = hasTitle && hasAuthor && hasDate;
+  let missingFields: string[] = [];
+  if (!isValid) {
+    !hasTitle && missingFields.push("title");
+    !hasAuthor && missingFields.push("author");
+    !hasDate && missingFields.push("date");
+    throw new Error(
+      `${
+        hasTitle ? `"${frontMatter.title}"` : "A blog post"
+      } was missing this required front matter: [${missingFields.join(", ")}].`
+    );
+  }
+}
+
+/**
+ * Builds the front matter of all posts into a `PostFrontMatterCollection`
+ * as a JSON string.
+ */
+export function buildFrontMatter(): string {
+  if (!fs.existsSync(POSTS_SOURCE_DIR)) {
+    return "";
+  }
+  const frontMatterCollection = fs
+    .readdirSync(POSTS_SOURCE_DIR)
+    .reduce((postMetadata: PostFrontMatterCollection, filename: string) => {
+      if (!isMarkdownFilename(filename)) {
+        return postMetadata;
+      }
+      /**
+       * Get file contents
+       */
+      const filepath = path.join(POSTS_SOURCE_DIR, filename);
+      const fileContents = fs.readFileSync(filepath).toString("utf-8");
+
+      /**
+       * Check for front matter and extract it
+       */
+      if (!/---/.test(fileContents)) {
+        console.error("Blog post has no front matter! Skipping cache!");
+        return postMetadata;
+      }
+      let frontMatter!: PostFrontMatter;
+      try {
+        frontMatter = fm<PostFrontMatter>(fileContents).attributes;
```

Comment:
This one was also in #93 . We can use `mdx-bundler` to parse the front matter, but it has to compile the entire post to get it, components, files, everything. We only use `front-matter` to make the top level JSON array that contains the front matter for every post for the `blog.tsx` route (the one that just lists out title, author, and date for every post starting with most recent). So we're not concerned with the blog contents for this situation anyway, just the front matter. Seemed like it was right to use the package made for this specific scenario rather than use `mdx-bundler` for only a small portion of its capability and a lot of wasted data and potentially wasted compile time.

It kind of comes down to how we want to think about this. Should we make code tailored to Remix Austin, a super small website? Or do we want this code to possibly serve as the start of a stack or npm package for others, who could potentially have large blogs? If it's the latter case, I would argue we should use a more specific package to get front matter because someone could potentially have a lot of blog posts.

As for the `/---/.test`, that part doesn't really have anything to do with `mdx-bundler` or `front-matter`. That was something I wrote so that if a person is guest writing a blog post for the site and doesn't have any front matter at all, there's something to check for that and log an error. It could also be useful for debugging in general if a post keeps skipping the cache.